go/service/gregor_test.go

```diff
@@ -178,3 +188,266 @@ func TestShowTrackerPopupMessage(t *testing.T) {
+		refMsgID := refMsg.Metadata().MsgID()
+
+		if !bytes.Equal(msgID.Bytes(), refMsgID.Bytes()) {
+			t.Fatalf("message IDs do not match, %s != %s, source: %s", msgID, refMsgID, source)
+		}
+	}
+}
+
+func TestSyncFresh(t *testing.T) {
+	tc := libkb.SetupTest(t, "gregor", 2)
+	defer tc.Cleanup()
+	tc.G.SetService()
+
+	// Set up client and server
+	h, server, uid := setupSyncTests(t, tc)
+
+	//Consume a bunch of messages to the server, and we'll sync them down
+	const numMsgs = 20
+	var refMsgs []gregor.InBandMessage
+	for i := 0; i < numMsgs; i++ {
+		msgid := newMsgID()
+		msg := server.newIbm(uid, msgid)
+		refMsgs = append(refMsgs, msg.ToInBandMessage())
+		server.ConsumeMessage(context.TODO(), msg)
+	}
+
+	// Sync messages down and see if we get 20
+	replayedMessages, consumedMessages, err := h.serverSync(context.TODO(), server)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	checkMessages(t, "replayed messages", replayedMessages, refMsgs)
+	checkMessages(t, "consumed messages", consumedMessages, refMsgs)
+}
+
+func TestSyncNonFresh(t *testing.T) {
+	tc := libkb.SetupTest(t, "gregor", 2)
+	defer tc.Cleanup()
+	tc.G.SetService()
+
+	// Set up client and server
+	h, server, uid := setupSyncTests(t, tc)
+
+	//Consume a bunch of messages to the server, and we'll sync them down
+	const numMsgs = 6
+	const msgLimit = numMsgs / 2
+	var refMsgs []gregor.InBandMessage
+	for i := 0; i < numMsgs; i++ {
+		msgid := newMsgID()
+		msg := server.newIbm(uid, msgid)
+		server.ConsumeMessage(context.TODO(), msg)
+		if i < msgLimit {
+			h.BroadcastMessage(context.TODO(), msg)
+			// We end up picking up the last one in the sync, since it's
+			// CTime is equal to when we start the sync, so just add it
+			if i == msgLimit-1 {
+				refMsgs = append(refMsgs, msg.ToInBandMessage())
+			}
+		} else {
+			refMsgs = append(refMsgs, msg.ToInBandMessage())
+		}
+	}
+
+	// Turn off fresh sync
+	h.freshSync = false
+
+	// We should only get half of the messages on a non-fresh sync
+	replayedMessages, consumedMessages, err := h.serverSync(context.TODO(), server)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	checkMessages(t, "replayed messages", replayedMessages, refMsgs)
+	checkMessages(t, "consumed messages", consumedMessages, refMsgs)
+}
+
+func TestSyncSaveRestoreFresh(t *testing.T) {
+	tc := libkb.SetupTest(t, "gregor", 2)
+	defer tc.Cleanup()
+	tc.G.SetService()
+
+	// Set up client and server
+	h, server, uid := setupSyncTests(t, tc)
+
+	//Consume a bunch of messages to the server, and we'll sync them down
+	const numMsgs = 6
+	const msgLimit = numMsgs / 2
+	var refReplayMsgs, refConsumeMsgs []gregor.InBandMessage
+	for i := 0; i < numMsgs; i++ {
+		msgid := newMsgID()
+		msg := server.newIbm(uid, msgid)
+		server.ConsumeMessage(context.TODO(), msg)
+		if i < msgLimit {
+			h.BroadcastMessage(context.TODO(), msg)
+			// We end up picking up the last one in the sync, since it's
+			// CTime is equal to when we start the sync, so just add it
+			if i == msgLimit-1 {
```

Comment:
what if there were multiple messages set at the last CTime. Is this sufficiently general?